src/h5web/vis-packs/core/heatmap/Mesh.tsx

```diff
@@ -1,61 +1,129 @@
-    rows,
-    cols,
-    values,
-    domain,
-    scaleType,
-    colorMap
-  );
+  const dataTexture = useMemo(() => {
+    const valuesArr = Float32Array.from(values);
+    return new DataTexture(valuesArr, cols, rows, RedFormat, FloatType);
+  }, [cols, rows, values]);
+
+  const colorMapTexture = useMemo(() => {
+    const interpolator = INTERPOLATORS[colorMap];
 
-  const material = useMemo(() => {
-    return (
-      textureData &&
-      new MeshBasicMaterial({
-        map: new DataTexture(textureData, cols, rows, RGBFormat),
+    const colors = Uint8Array.from(
+      Array.from({ length: CMAP_SIZE }).flatMap((_, i) => {
+        const { r, g, b } = rgb(interpolator(i / (CMAP_SIZE - 1)));
+        return [r, g, b];
       })
     );
-  }, [cols, rows, textureData]);
+
+    return new DataTexture(colors, CMAP_SIZE, 1, RGBFormat, UnsignedByteType);
+  }, [colorMap]);
+
+  const shader = {
+    uniforms: {
+      data: { value: dataTexture },
+      colorMap: { value: colorMapTexture },
+      scaleType: { value: CMAP_NORM[scaleType] },
+      min: { value: scaledDomain[0] },
+      oneOverRange: { value: 1 / (scaledDomain[1] - scaledDomain[0]) },
+    },
+    vertexShader: `
+      varying vec2 coords;
+
+      void main() {
+        coords = uv;
+        gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );
+      }
+    `,
+    fragmentShader: `
+      uniform sampler2D data;
+      uniform sampler2D colorMap;
+      uniform int scaleType;
+      uniform float min;
+      uniform float oneOverRange;
+
+      const float oneOverLog10 = 0.43429448190325176;
+      const vec4 nanColor = vec4(255, 255, 255, 1);
+
+      varying vec2 coords;
+
+      float symlog(float x) {
+        return sign(x) * log(1. + abs(x)) * oneOverLog10;
+      }
+
+      float scale(float value) {
+        if (scaleType == 1) {
+          return oneOverRange * (log(value) * oneOverLog10 - min); // Log
+        }
+
+        if (scaleType == 2) {
+          return oneOverRange * (symlog(value) - min); // SymLog
+        }
+
+        return oneOverRange * (value - min); // Linear
+      }
+
+      void main() {
+        float value = texture(data, coords).r;
+
+        if (scaleType == 1 && value <= 0.) {
+          gl_FragColor = nanColor;
+        } else {
+          gl_FragColor = texture(colorMap, vec2(scale(value), 0.5));
+        }
+      }
+    `,
+  };
+
+  const { size } = useThree();
+  const { width, height } = size;
 
   return (
-    <>
-      {material && (
-        <mesh material={material}>
-          <planeGeometry attach="geometry" args={[width, height]} />
-        </mesh>
-      )}
-      {showLoader && (
-        <Html
-          className={styles.textureLoader}
-          data-visible={loading || undefined}
-        />
-      )}
-    </>
+    <mesh scale={[width, height, 1]}>
+      <planeGeometry attach="geometry" args={[1, 1]} />
+      <shaderMaterial attach="material" args={[shader]} />
+    </mesh>
```

Comment:
Why not keeping `width, height` on the geometry ?
```js
   <mesh>
      <planeGeometry attach="geometry" args={[width, height]} />
      <shaderMaterial attach="material" args={[shader]} />
    </mesh>
```

Reply:
If I do, the `uv` values are not between `0` and `1` but, I think, between `0` and `width` or `height`, which doesn't work well with the `texture()` function -- we get near-solid heatmap with only the "highest" colour in the colour map.